pennylane/templates/embeddings/iqp.py

```diff
@@ -0,0 +1,249 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Contains the ``IQPEmbedding`` template.
+"""
+# pylint: disable-msg=too-many-branches,too-many-arguments,protected-access
+from collections import Iterable
+from pennylane.templates.decorator import template
+from pennylane.ops import RZ, CNOT, Hadamard
+from pennylane.templates import broadcast
+from pennylane.templates.utils import (
+    check_shape,
+    check_wires,
+    check_type,
+    get_shape,
+    check_no_variable,
+)
+
+
+@template
+def zz(parameter, wires):
+    """Template for decomposition of ZZ coupling.
+
+    Args:
+        parameter (float): parameter of z rotation
+        wires (list[int]): qubit indices that the template acts on
+    """
+    CNOT(wires=wires)
+    RZ(2 * parameter, wires=wires[0])
+    CNOT(wires=wires)
+
+
+@template
+def IQPEmbedding(features, wires, n_repeats=1, pattern=None):
+    r"""
+    Encodes :math:`n` features into :math:`n` qubits using diagonal gates of an IQP circuit. The embedding
+    has been proposed by `Havlicek et al. (2018) <https://arxiv.org/pdf/1804.11326.pdf>`_. It and can be repeated.
+
+    .. warning::
+
+        ``IQPEmbedding`` calls a circuit that involves non-trivial classical processing of the
+        features. The ``features`` argument is therefore **not differentiable** when using the template, and
+        gradients with respect to the features cannot be computed by PennyLane.
+
+    An IQP circuit is a quantum circuit of a block of Hadamards, followed by a block of gates that are
+    diagonal in the computational basis. Here, the diagonal gates are single-qubit ``RZ`` rotations, applied to each
+    qubit and encoding the :math:`n` features, followed by two-qubit ZZ entanglers,
+    :math:`e^{-i x_i x_j \sigma_z \otimes \sigma_z}`. The entangler applied to wires ``(i, j)`` encodes the
+    product of features ``features[i]*features[j]``. The pattern in which the entanglers are applied is either the
+    default, or a custom pattern:
+
+    * If ``pattern`` is not specified, the default pattern will be used, in which the entangling gates connect all
+      pairs of neighbours:
+
+      |
+
+      .. figure:: ../../_static/templates/embeddings/iqp.png
+          :align: center
+          :width: 50%
+          :target: javascript:void(0);
+
+      |
+
+    * Else, ``pattern`` is a list of wire pairs ``[[a, b], [c, d],...]``, applying the entangler
+      on wires ``[a, b]``, ``[c, d]``, etc. For example, ``pattern = [[0, 1], [1, 2]]`` produces
+      the following entangler pattern:
```

Comment:
Ah, so this is why the custom pattern was needed 😆

Reply:
Yes, neat, hey?